servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingHttpRequesterFilter.java

```diff
@@ -210,8 +219,32 @@ public Completable apply(final int count, final Throwable t) {
             }
 
             Completable applyRetryCallbacks(final Completable completable, final int retryCount, final Throwable t) {
-                return retryCallbacks == null ? completable :
-                        completable.beforeOnComplete(() -> retryCallbacks.beforeRetry(retryCount, requestMetaData, t));
+                Completable result = (retryCallbacks == null ? completable :
+                        completable.beforeOnComplete(() -> retryCallbacks.beforeRetry(retryCount, requestMetaData, t)));
+                if (returnFailedResponses) {
+                    if (t instanceof HttpResponseException &&
+                            ((HttpResponseException) t).metaData() instanceof StreamingHttpResponse) {
+                        StreamingHttpResponse response = (StreamingHttpResponse) ((HttpResponseException) t).metaData();
+                        // If we succeed, we need to drain the response body before we continue. If we fail we want to
+                        // surface the original exception and don't worry about draining since it will be returned to
+                        // the user.
+                        result = result.onErrorMap(backoffError -> ThrowableUtils.addSuppressed(t, backoffError))
+                                // If we get cancelled we also need to drain the message body as there is no guarantee
+                                // we'll ever receive a completion event, error or success.
+                                .beforeCancel(() -> drain(response).subscribe())
+                                .concat(drain(response));
+                    } else if (LOGGER.isDebugEnabled()) {
+                        if (!(t instanceof HttpResponseException)) {
+                            LOGGER.debug("Couldn't unpack response due to unexpected dynamic types. Required " +
+                                    "exception of type HttpResponseException, found {}", t.getClass());
```

Comment:
Do we need to log in case t is not `HttpResponseException`? Any other exception type will fall under this condition and produce logging. Users may come to us worried about it.

+1 for logging when the exception type is `HttpResponseException` but metaData is not `StreamingHttpResponse`. It might be even better to increase this to info/warn and add an assertion. 
Should we log here or inside `single.onErrorResume(HttpResponseException.class, ...)` or in both places?